### PR TITLE
CT-1386 - backend assing test when string is passed in request

### PR DESCRIPTION
**Pull request by vojtabiberle**
Jira: [CT-1386](https://keboola.atlassian.net/browse/CT-1386)
Connection PR: https://github.com/keboola/connection/pull/4992

Before asking for review make sure that:

## Checklist

- [x] New client method(s) has tests
- [x] Apiary file is updated
- [x] You declared if there is a BC break or not (will affect next release of a client)

## Release

- [x] I gave the PR a proper label:
    * major (BC break)
    * minor (new feature)
    * patch (backwards compatible fix)
    * no release (just test changes)

---

[CT-1386]: https://keboola.atlassian.net/browse/CT-1386?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: tests/ProjectStorageBackendTest.php [hunk 1]
```diff
@@ -90,6 +92,47 @@ public function testProjectStorageAssignBackend(string $backendName): void
...
+        $organization = $this->client->createOrganization($this->testMaintainerId, [
+            'name' => $name,
+        ]);
+
+        $project = $this->client->createProject($organization['id'], [
+            'name' => 'My test',
+            'dataRetentionTimeInDays' => 1,
+        ]);
+
+        $apiPostMethod = new ReflectionMethod(ManageClient::class, 'apiPost');
```
Comment (romanbracinik) [inline tests/ProjectStorageBackendTest.php:107]: Tu by  som sa zamyslel ci si toho klienta nevytvorit nejak rucne bokom napriklad tak ako to je tu https://github.com/keboola/kbc-manage-api-php-client/blob/86166a67f3da8b861a47b5a782b1f2d8b0c7f4f3/tests/CommonTest.php#L51-L79, ale obecne proti tomu nic nemam, takto by som to riesil ja, plus by si potom nemusel mat ten phpstan ignor na riadku 114 a 124

File: tests/ProjectStorageBackendTest.php [hunk 2]
```diff
@@ -90,6 +92,47 @@ public function testProjectStorageAssignBackend(string $backendName): void
...
+
+        try {
+            $apiPostMethod->invoke($this->client, '/manage/projects/' . $project['id'] . '/storage-backend', [
+                'storageBackendId' => '666',
+            ]);
+            // @phpstan-ignore-next-line calling by `invoke` is not recognized by PHPStan as this is dynamic call by reflection
+        } catch (ClientException $e) {
+            // backend storage '666' not found
+            // here we're only testing the format of `storageBackendId` so we don't care about the existence of backend storage
+            $this->assertEquals(500, $e->getCode());
```
Comment (romanbracinik) [inline tests/ProjectStorageBackendTest.php:128]: Pride mi, ze by to nemalo vratit Internal err, v pripade ze tam davas idcko ktore neexistuje. Druha vec je ze tento test https://github.com/keboola/kbc-manage-api-php-client/blob/20e8e1cd78bd40ceb3eb7830a34ab86233a6c04a/tests/StorageBackendTest.php#L176 bohuzial po sebe nemaze zaregistrovane backendy. Takze ked si vylistujes vsetky backendy https://keboolamanagementapi.docs.apiary.io/#reference/super-storage-backends-management/storage-backend-collection/list-backends napr na https://connection.e2e.us-east-1.aws.keboola.dev/ tak tam to Idcko teoreticky moze existovat. 

Takze test by si mal idealne zistit posledne idecko a testovat +1. Manage zatial nebezi paralelne takze by to asi slo. Pripadne ma napada, ze by to slo este otestovat tak, ze si zalozim backend rovno zmazem a to IDcko pouzijem na tento test.

File: tests/ProjectStorageBackendTest.php [hunk 3]
```diff
@@ -90,6 +92,47 @@ public function testProjectStorageAssignBackend(string $backendName): void
...
+        ]);
+
+        $apiPostMethod = new ReflectionMethod(ManageClient::class, 'apiPost');
+        $apiPostMethod->setAccessible(true);
+
+        try {
+            $apiPostMethod->invoke($this->client, '/manage/projects/' . $project['id'] . '/storage-backend', [
+                'storageBackendId' => 'non-numeric',
+            ]);
+            // @phpstan-ignore-next-line calling by `invoke` is not recognized by PHPStan as this is dynamic call by reflection
```
[the remark this answers is not in the document]
Comment (vojtabiberle) [reply tests/ProjectStorageBackendTest.php:114]: :100:

File: tests/ProjectStorageBackendTest.php [hunk 4]
```diff
@@ -90,6 +92,47 @@ public function testProjectStorageAssignBackend(string $backendName): void
...
+        } catch (ClientException $e) {
+            $this->assertEquals(400, $e->getCode());
+            $this->assertEquals('storageBackendId must be an integer.', $e->getMessage());
+        }
+
+        try {
+            $apiPostMethod->invoke($this->client, '/manage/projects/' . $project['id'] . '/storage-backend', [
+                'storageBackendId' => '666',
+            ]);
+            // @phpstan-ignore-next-line calling by `invoke` is not recognized by PHPStan as this is dynamic call by reflection
```
[the remark this answers is not in the document]
Comment (vojtabiberle) [reply tests/ProjectStorageBackendTest.php:124]: :100:

File: tests/ProjectStorageBackendTest.php (hunk 1 above)
Comment (vojtabiberle) [reply tests/ProjectStorageBackendTest.php:107]: to je velmi dobrý nápad, jen se nesmí změnit nastavení, které očekáváme od klienta, ale to asi nebude časté

File: tests/ProjectStorageBackendTest.php (hunk 2 above)
Comment (vojtabiberle) [reply tests/ProjectStorageBackendTest.php:128]: Vrací to `BackendNotFoundException`: https://github.com/keboola/connection/blob/3f73a3882fe203af129dc7f22dedfec230ceb738/legacy-app/application/src/Storage/Service/Backend/Assign/BackendAssign.php#L140

Kterou následně konvertuje `ExceptionConverter` na 500: https://github.com/keboola/connection/blob/3f73a3882fe203af129dc7f22dedfec230ceb738/legacy-app/application/src/Common/Exception/JsonExceptionConverter/ExceptionConverter.php#L65

Ale s tím ID+1 máš pravdu.

File: tests/ProjectStorageBackendTest.php [hunk 5]
```diff
@@ -90,6 +92,65 @@ public function testProjectStorageAssignBackend(string $backendName): void
...
+        try {
+            $requestOptions['body'] = '{"storageBackendId": "non-numeric"}';
+            $client->post('/manage/projects/' . $project['id'] . '/storage-backend', $requestOptions);
+            $this->fail('Should fail with 400');
+        } catch (GuzzleException $e) {
+            $this->assertEquals(400, $e->getCode());
+            $this->assertStringContainsString('storageBackendId must be an integer.', $e->getMessage());
+        }
+
+        $this->client->removeProjectStorageBackend($project['id'], $project['backends']['snowflake']['id']);
```
Comment (romanbracinik) [inline tests/ProjectStorageBackendTest.php:129]: Toto ma este dve nevyhody, keby sme sa rozhodli z nejakeho dovodu ze nam budu bezat manage testy nad inym backendom tak to musime upravit. Alebo teoreticky ak na testingu nema ten maintainer default backend tak to bude null a spadne to.

Co si pre ten test zalozi vlastneho mainstainera? Je to len metadatova operacia:
-  zalozis noveho maintainera a nenastavis mu ziadny default backend
- potom nemusis mazat projectbackend, ale len ho priradis.

pred kazdy testom sa automaciky premazavaju maintainery ak to neieje ten tvoj hlavny https://github.com/keboola/kbc-manage-api-php-client/blob/c5decb33af65a73aec461b2bd61360dab69b37a5/tests/ClientTestCase.php#L194-L207 takze nemusis robit ziadny cleaup.

**Review by romanbracinik — COMMENTED** [review]
mam ktomu uz len dve veci a pak si myslim ze to bude cajk ;)

File: tests/ProjectStorageBackendTest.php (hunk 5 above)
Comment (vojtabiberle) [reply tests/ProjectStorageBackendTest.php:129]: good!

File: tests/ProjectStorageBackendTest.php [hunk 6]
```diff
@@ -90,6 +92,65 @@ public function testProjectStorageAssignBackend(string $backendName): void
...
+        // ensure, that backend ID is passed as string into body
+        $requestOptions['body'] = '{"storageBackendId": "'.$backend['id'].'"}';
+        $response = $client->post('/manage/projects/' . $project['id'] . '/storage-backend', $requestOptions);
+        $this->assertEquals(200, $response->getStatusCode());
+
+        $this->client->deleteProject($project['id']);
+        $this->client->purgeDeletedProject($project['id']);
+    }
+
+    public function getBackendCreateOptions(): array
```
[the remark this answers is not in the document]
Comment (vojtabiberle) [reply tests/ProjectStorageBackendTest.php:141]: Ano. Hodil jsem to do traitu. 

Ale moc nemám rád tohle znovupoužití ve více testech. Až to někdo bude v jednom testu jinak a v jiném nechat stejně, opět muset vymejšlet, jak to smysluplně roztrhávat. A většinou to vede k nějakému parametru, než k jednoduchému zapsání metody zpět do testu.

File: tests/ProjectStorageBackendTest.php (hunk 5 above)
Comment (vojtabiberle) [reply tests/ProjectStorageBackendTest.php]: done

**Review by romanbracinik — APPROVED** [review]
🙆‍♂️  este to prosim ta vyskusaj pridal som tam fa54c02b847d44dfa717ab3c346b4272495e305e mazanie toho vytvoreneho backendu na koniec aby nam to na testingu neboptnalo. Jak upravujem teraz tie fronty tak som si rozbil connection

File: tests/ProjectStorageBackendTest.php [hunk 7]
```diff
@@ -90,6 +94,56 @@ public function testProjectStorageAssignBackend(string $backendName): void
...
+
+        $backend = $this->client->createStorageBackend($this->getSnowflakeBackendCreateOptions());
+
+        // ensure, that backend ID is passed as string into body
+        $requestOptions['body'] = '{"storageBackendId": "'.$backend['id'].'"}';
+        $response = $client->post('/manage/projects/' . $project['id'] . '/storage-backend', $requestOptions);
+        $this->assertEquals(200, $response->getStatusCode());
+
+        $this->client->deleteProject($project['id']);
+        $this->waitForProjectPurge($project['id']);
```
Comment (romanbracinik) [inline tests/ProjectStorageBackendTest.php:142]: ten purge sa da [volat](https://keboolamanagementapi.docs.apiary.io/#reference/deleted-projects/purge-deleted-project/purge-deleted-project) aj s parametrom `ignoreBackendErrors`. Takze rychly fix je pridat do `waitForProjectPurge` parameter `ignoreBackendErrors = false` a v tej metode vo vnutri preposlat ten parameter na https://github.com/keboola/kbc-manage-api-php-client/blob/c5decb33af65a73aec461b2bd61360dab69b37a5/tests/ClientTestCase.php#L344, tu metodu `waitForProjectPurge` pouzivame len v testoch tak by to asi slo ako quick fix, ale chcelo by to opravit asi v connection, ale imho to moze is tako dalsi task

**Review by romanbracinik — COMMENTED** [review]
Hlavny problem je, ze ak mas novy file storage tak tam este neexistuju ziadne foldre a volanie https://github.com/keboola/connection/blob/3f73a3882fe203af129dc7f22dedfec230ceb738/legacy-app/application/modules/core/services/FileStorage/Adapter/S3Adapter.php#L275 vrati null (pretoze https://docs.aws.amazon.com/AmazonS3/latest/API/API_CommonPrefix.html )a spadne pak ten foreach https://github.com/keboola/connection/blob/3f73a3882fe203af129dc7f22dedfec230ceb738/legacy-app/application/modules/core/services/FileStorage/Adapter/S3Adapter.php#L277 

My sme na to asi nenarazili, pretoze vatsinou sa ten projekt purguje ked mal nejake uploady, alebo ked to padalo tak sme to purgli s ignoreBackendError. 

Tak to necham na teba ci to teraz fixnes len v testoch a pridas ten `ignoreBackendError` a zalozis si task na fix v connection. Alebo to rovno fixnes v connection

File: tests/ProjectStorageBackendTest.php (hunk 7 above)
Comment (vojtabiberle) [reply tests/ProjectStorageBackendTest.php:142]: opraveno v connection: https://github.com/keboola/connection/pull/4992/commits/1c3345099d9385421c72e0bf4782519282800420

testy fungují

**Pull request merged**